### ISSUE TITLE
Read and Write are not symmetric

### ISSUE THREAD

**Issue by HarelM** [NetTopologySuite/NetTopologySuite.IO.GPX#14]
I would expect the read and write to be symmetric.
The following `Write` method doesn't have a `Read` counterpart:
https://github.com/NetTopologySuite/NetTopologySuite.IO.GPX/blob/24240b5610cc625050f1e6eb979dba0fadae34cc/src/NetTopologySuite.IO.GPX/GpxWriter.cs#L72
Meaning I can't write and then read to the same objects...

Also `GpxVisitorBase` is not something I would expect to see in a `Read` function - I have no idea what to put there - should I inherit? is the only option is `NetTopologySuiteFeatureBuilderGpxVisitor` which doesn't seem like what I need...

**Comment by airbreather** [comment]
> Also `GpxVisitorBase` is not something I would expect to see in a `Read` function - I have no idea what to put there [...]

`Read` doesn't buffer more than a single data element (`GpxMetadata`, `GpxWaypoint`, `GpxRoute`, `GpxTrack`, or outermost extension) at a time.  To make that work, given the fact that there are at least 5 different kinds of top-level children of the `gpx` element, I saw two different approaches:
- return something like `IEnumerable<X>`, where `X` is a discriminated union of all the different kinds of events that we can see as we go through the file, or
- have the caller give us callbacks that we can invoke when we see the next element in the stream

I decided to go the second approach.  Digging deeper into that approach, there were a few different ways I could have gone with it... this weird visitor-pattern thing is kinda wacky, but all the alternatives that come to mind don't have the flexibility to do some of the other things I intend to do, without just bringing this same pattern back at a weird level.

In any case, the overall idea is that `Read` is, in fact, the counterpart to that `Write` method that you linked to.  For example, whereas `Write` models a stream of waypoints as `IEnumerable<GpxWaypoint>`, `Read` models it as a stream of calls to the `visitor.VisitWaypoint(GpxWaypoint)` method.

> [...] should I inherit?

Yeah, at this point if you don't want the tuple of `(GpxMetadata metadata, Feature[] features, object extensions)` that `ReadFeatures` will give you, then your only option while reading is to make a subclass of `GpxVisitorBase` that does whatever you want it to do as it sees each individual element of the original file.

All that said, it wouldn't be all that hard for me to make a simple visitor to use when you don't mind buffering, and then wrap it like `ReadFeatures` does.

**Comment by airbreather** [comment]
At this point, I'm going to interpret this issue as:

    Add something like `public static (GpxMetadata metadata, GpxWaypoint[] waypoints, GpxRoute[] routes, GpxTrack[] tracks, object extensions) ReadAll(XmlReader reader, GpxReaderSettings settings)`

and commit to doing that for the next minor release increment.

**Comment by airbreather** [comment]
Of course, it would make more sense for it to actually return something like what #12 suggests.

**Comment by HarelM** [comment]
Sounds about right :-)
